### ISSUE TITLE
Azure KeyVault Error: Provider produced inconsistent result after apply

### ISSUE THREAD

**Issue by GregBillings** [hashicorp/terraform-provider-azurerm#11059]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.14.8
AzureRM: terraform-provider-azurerm_v2.52.0_x5

### Affected Resource(s)

* `azurerm_key_vault_secret `

### Terraform Configuration Files

```hcl
resource "azurerm_key_vault_secret" "mysecretvalue" {
  name         = "secretvaluename"
  value        = var.some_value_from_var
  key_vault_id = data.terraform_remote_state.remote_terraform_cloud_state.outputs.key_vault_id
}
```

### Debug Output

Error: Provider produced inconsistent result after apply

When applying changes to azurerm_key_vault_secret.mysecretvalue,
provider "registry.terraform.io/hashicorp/azurerm" produced an unexpected new
value: Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.

### Expected Behaviour

secret added to the keyvault

### Actual Behaviour

The keyvault secret was added to the keyvault with the correct value, but the terraform apply failed with the error above. When re-running again, the new error is that the value already exists but isn't tracked in the terraform state

### Steps to Reproduce

1. Create a terraform script that creates an azure keyvault and then outputs the ID as an output variable
2. Create another terraform script with a different remote state that pulls in the first remote state via:
```hcl
data "terraform_remote_state" "remotestate" {
  backend = "remote"

  config = {
    organization = "my-org"
    workspaces = {
      name = "first-remote-state"
    }
  }
}
```
3. Attempt to create a new keyvault secret using the id from the output of the first remote state:
```hcl
resource "azurerm_key_vault_secret" "mysecretvalue" {
  name         = "MySecretValue"
  value        = var.some_value_from_var
  key_vault_id = data.terraform_remote_state.remotestate.outputs.key_vault_id
}
```

### References

* #10227

**Comment by dkirrane** [comment]
I had raised this previously here https://github.com/terraform-providers/terraform-provider-azurerm/issues/10227. 
Still seeing it with `Terraform v0.14.9` & `hashicorp/azurerm 2.56.0`

**Comment by elthanor** [comment]
Have the same issue with azurerm v2.71
terraform 1.0.4

**Comment by sinbai** [comment]
@GregBillings I am investigating this issue, but I can't repro it by following the steps below. Could you help confirm whether the following steps can repro this issue on your side? In addition, I would like to confirm whether you have done other operations before doing the following?

**Steps to Reproduce**
1. Creates an azure keyvault and then outputs the ID as an output variable via:
    terraform {
      required_providers {
        azurerm = {
          source = "hashicorp/azurerm"
         version = "2.52.0"
        }
      }
    
       backend "remote" {
       organization = "my-org-elena"
        workspaces {
          name = "ReproBug"
        }
     }
    }
    
    provider "azurerm" {
      features {}
    }
    
    resource "azurerm_resource_group" "test" {
      name     = "myTFResourceGroup"
      location = "westus2"
    }
    
    resource "azurerm_key_vault" "test" {
      name                       = "acctestkv-elena05"
      location                   = azurerm_resource_group.test.location
      resource_group_name        = azurerm_resource_group.test.name
      tenant_id                  = "XXXXXXXXX"
      sku_name                   = "standard"
      soft_delete_retention_days = 7
    
      access_policy {
        tenant_id = "XXXXXXX"
        object_id = "XXXXXXX"
    
        key_permissions = [
          "Get",
         "Delete",
        ]
    
        secret_permissions = [
          "Get",
          "Delete",
          "List",
          "Purge",
          "Recover",
          "Set",
        ]
      }
    
      tags = {
        environment = "Production"
      }
    }
    
    output "key_vault_id" {
      value = azurerm_key_vault.test.id
    }

2. create a new keyvault secret using the id from the output of the first remote state via:
    terraform {
      required_providers {
        azurerm = {
          source = "hashicorp/azurerm"
         version = "2.52.0"
        }
      }
    
       backend "remote" {
       organization = "my-org-elena"
        workspaces {
          name = "ReproBug1"
        }
     }
    }
    
    provider "azurerm" {
      features {}
    }
    
    data "terraform_remote_state" "remotestate" {
      backend = "remote"
    
      config = {
        organization = "my-org-elena"
        workspaces = {
          name = "ReproBug"
        }
      }
    }
    
    resource "azurerm_key_vault_secret" "mysecretvalue" {
      name         = "MySecretValue2"
      value        = "test"
      key_vault_id = data.terraform_remote_state.remotestate.outputs.key_vault_id
    }

**My Terraform (and AzureRM Provider) Version**
<img width="1571" alt="Untitled" src="https://user-images.githubusercontent.com/39109137/131312604-8c8d6581-0443-4022-9459-ee2f93c44cd0.png">

**Comment by sinbai** [comment]
> Have the same issue with azurerm v2.71
> terraform 1.0.4
> 
> deleted the remote state and recreated from scratch and it worked fine. I previously upgraded from 2.70 -> 2.71 and terraform 1.0.1 -> 1.0.4

@elthanor  I can't repro this issue, below is my repro steps. Could you provide more details to help me repro this issue?
1.Create two new terraform cloud workspaces(ReproBug1 and ReproBug2).
2.Create keyvault with azurerm v2.70 and terraform 1.0.1 on ReproBug1 workspace.
3.Update to azurerm provider to v2.7.1 and terraform to 1.0.4, add a new secret to the existing keyvault on workspace ReproBug2. 

Note: Remote state is managed by Terraform Cloud, issue can't be reproduced when Execution Mode is remote or local.

**Comment by mybayern1974** [comment]
I would like to Re sinbai's finding that I cannot repro this issue either. The app I used is TF v1.0.5 and AzureRM provider v2.74 (this is the latest version when I'm typing now).  What I tried is creating a KV + KVSec first, and then use that created KV as a data source and create another KVSec. All the above was done by TF and all of them worked well w/o getting any command line error msg. I admit I just tried them w/ local state rather than remote state.

With noticing there are 20+ thump up to this issue, I do believe people ran into issues as described here. While to serve my or sinbai's troubleshooting, can anyone here provide more contexts in terms of step-by-step-repro-this-issue? 

In addition, it would be helpful if below info can be provided:
1. Does this issue only repro when using remote state rather than local state?
2. Does this issue stable repro or happen intermittently?
3. For people who ran into this issue, do you use TF to manage all things (KV, KVSec, etc.) rather than using any other client tools (portal, CLI, etc) to co-manage resources?
4. For people who ran into this issue, is it possible there were someone-else/some-other-client-tooling manipulating the same resource (KV) at the same time when you used TF to manage that? If so, might this symptom be a result of conflict-manipulation-against-the-same-KV?

**Comment by viper4u** [comment]
@mybayern1974 
Terraform 1.0.0
azurerm 2.71.0

1) I use backend "azurerm" 
2) The problem occurs in ~70% of attempts
3) KeyVault has been created separately with the local state and I don't have this issue with the local state
4) We don't use KeyVault at the same moment

**Comment by doug-papenthien-by** [comment]
When I ran into this issue, I was using a `data` resource to look up information about another key vault that was created outside of my terraform scripts.  I was using the resource ID returned by the `data` source to create additional `azurerm_key_vault_secret` resources in that vault.

**Comment by mybayern1974** [comment]
@viper4u , I still could not repro after using remote states (use backend "azurerm") and using TF/AzureRM version you specified 😕. With seeing you mentioned 70% repro rate, I ran things 10 times and all of them succeeded. 

Below are my repro steps with .tf config

1. Create a resource group on portal
2. Create a KV by TF with using backend=azurerm and TF ver=1.0.0 and AzureRM ver=2.71.0. Below is my config
===kv.tf===
```
terraform {
  backend "azurerm" {
    resource_group_name = "..."
    storage_account_name = "..."
    container_name = "..."
    key = "terraform1.tfstate"
  }
  required_providers {
    azurerm = {
      source = "hashicorp/azurerm"
      version = "2.71"
    }
  }
}

provider "azurerm" {
  features {}
}

data "azurerm_client_config" "current" {}

data "azurerm_resource_group" "test" {
  name     = "..."
}

resource "azurerm_key_vault" "test" {
  name                = "..."
  location            = data.azurerm_resource_group.test.location
  resource_group_name = data.azurerm_resource_group.test.name
  sku_name = "standard"
  tenant_id = data.azurerm_client_config.current.tenant_id

  access_policy {
    tenant_id = data.azurerm_client_config.current.tenant_id
    object_id = data.azurerm_client_config.current.object_id

    key_permissions = [
      "create",
      "get",
      "list",
      "purge"
    ]

    secret_permissions = [
      "set",
      "get",
      "delete",
      "purge",
      "recover",
      "list"
    ]
  }
}
```
===outputs.tf===
```
output "key_vault_id" {
  value = azurerm_key_vault.test.id
}
```
4. Execute `terraform apply` => A KV got provisioned
5. Create a KVSec by TF. Below is my .tf config
===kvsec.tf===
```
data "terraform_remote_state" "test" {
  backend = "azurerm"
  config = {
    storage_account_name = "..."
    container_name       = "..."
    resource_group_name  = "..."
    key                  = "terraform1.tfstate"
  }
}

resource "azurerm_key_vault_secret" "test" {
    name  = "..."
    value = "..."
    key_vault_id = data.terraform_remote_state.test.outputs.key_vault_id
}
```
6. Execute `terraform apply` => A KVSec got provisioned underneath the KV
7. Append below section to the `kvsec.tf`
```
resource "azurerm_key_vault_secret" "test2" {
    name  = "..."
    value = "..."
    key_vault_id = data.terraform_remote_state.test.outputs.key_vault_id
}
```
8. Execute `terraform apply` => Another KVSec got provisioned underneath the KV
10. Repeat step 7 - Step 8 for say 10 times => A bunch of KVSec got provisioned underneath the KV w/o seeing tf command line errors. 

@doug-papenthien-by , I also did similar things by using local state and follow your "_I was using the resource ID returned by the data source to create additional azurerm_key_vault_secret_". While I still could not repro.

@ all here, do you see any usage/config difference between yours and mine above?

**Comment by sinbai** [comment]
> @mybayern1974
> Terraform 1.0.0
> azurerm 2.71.0
> 
> 1. I use backend "azurerm"
> 2. The problem occurs in ~70% of attempts
> 3. KeyVault has been created separately with the local state and I don't have this issue with the local state
> 4. We don't use KeyVault at the same moment

@viper4u Thank you for your reply. Below are my repro steps. The issue can't be reproduced. Could you follow the stpes below to reproduce it? If not, could you provide the step-by-step to help me reproduce this issue?

**My Terraform (and AzureRM Provider) Version**
Terraform 1.0.0
azurerm 2.71.0

**Steps**
1. Create a resource group named `myTestResourceGroup` on Azure portal. 
2. Create a blob storage account named `teststorageaccount` on the Azure portal. (Stores the state as a Blob with the given Key within the Blob Container within the Blob Storage Account.)
3. Create a blob container named `blobcontainer` on the Azure portal. 
4. Create a folder named `ReproBug` on the local machine and add a file named "step1.tf" in this folder.
5. Add the following tfconfig in `step1.tf` to create a azurerm_key_vault(Authenticate using a SAS Token associated with the Storage Account) :
```
terraform {
  required_providers {
    azurerm = {
      source = "hashicorp/azurerm"
     version = "2.71.0"
    }
  }

   backend "azurerm" {
    storage_account_name = "teststorageaccount"
    container_name       = "blobcontainer"
    key                  = "prod.terraform.tfstate1"

    sas_token = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
  }
}


provider "azurerm" {
  features {}
}


resource "azurerm_key_vault" "test" {
  name                       = "acctestkv-elena0901"
  location                   = "westus2"
  resource_group_name        = "myTestResourceGroup"
  tenant_id                  = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
  sku_name                   = "standard"
  soft_delete_retention_days = 7

  access_policy {
    tenant_id = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
    object_id = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXX"

    key_permissions = [
      "Get",
	  "Delete",
    ]

    secret_permissions = [
      "Get",
      "Delete",
      "List",
      "Purge",
      "Recover",
      "Set",
    ]
  }

  tags = {
    environment = "Production"
  }
}

output "key_vault_id" {
  value = azurerm_key_vault.test.id
}
```
6. Run `terraform init`、`terraform plan`、 `terraform apply` to create azurerm_key_vault.
7. Add a file named "step2.tf" in `ReproBug` folder.
8. Add the following tfconfig in `step2.tf` to create a azurerm_key_vault_secret(Authenticate using a SAS Token associated with the Storage Account):
```
terraform {
  required_providers {
    azurerm = {
      source = "hashicorp/azurerm"
     version = "2.71.0"
    }
  }

   backend "azurerm" {
    storage_account_name = "teststorageaccount"
    container_name       = "blobcontainer"
    key                  = "prod.terraform.tfstate2"

    sas_token = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
  }
}

provider "azurerm" {
  features {}
}

data "terraform_remote_state" "remotestate" {
  backend = "azurerm"

  config = {
    storage_account_name = "teststorageaccount"
    container_name       = "blobcontainer"
    key                  = "prod.terraform.tfstate1"

    sas_token = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
  }
}

resource "azurerm_key_vault_secret" "mysecretvalue" {
  name         = "MySecretValue-0901"
  value        = "test"
  key_vault_id = data.terraform_remote_state.remotestate.outputs.key_vault_id
}

```

9. After successfully creating azurerm_key_vault, run `terraform init`、`terraform plan`、 `terraform apply` to create azurerm_key_vault_secret=>azurerm_key_vault_secret was created,no command line error msg 

10. Continue to add ten azurerm_key_vault_secrets with different `key` and different azurerm_key_vault_secret `name` to the azurerm_key_vault  which created in step 6.=>All of them worked well, There is no command line error msg in my test

**Comment by viper4u** [comment]
@sinbai @mybayern1974 
Sorry for the late answer 
I tried to reproduce it with a simple script but I haven't faced the problem.
I have a failed attempt with a trace log level. 
Where can I send it?

**Comment by sinbai** [comment]
> @sinbai @mybayern1974
> Sorry for the late answer
> I tried to reproduce it with a simple script but I haven't faced the problem.
> I have a failed attempt with a trace log level.
> Where can I send it?

@viper4u  Mail log to sinbai@hotmail.com, thank you. I will try to see if there are any findings in the log, but I think step-by-step-repro-this-issue is the key to investigating this issue. Thank you very much.

**Comment by viper4u** [comment]
@sinbai 
I've reproduced it with a simple script. 
I have an Azure KeyVault and that issue is reproduced each time with that one
How can I help you to with it?

**Comment by sinbai** [comment]
> @sinbai
> I've reproduced it with a simple script.
> I have an Azure KeyVault and that issue is reproduced each time with that one
> How can I help you to with it?

Can I access Azure KeyVault? And could you share your script (contains the access token) with me to reproduce it on my side?

**Comment by viper4u** [comment]
What did you mean by "access token"? I use [Service Principal auth](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/guides/service_principal_client_secret)

**Comment by viper4u** [comment]
I've sent all details to your mail

**Comment by sinbai** [comment]
> I've sent all details to your mail

Thank you. Per details provided, I cannot access the Azure KeyVault due to the access permissions.

**Comment by viper4u** [comment]
I've reproduced the issue with new keyvault
Context:

- Windows 10
- Terraform version: 1.0.6
- Service principle auth 

Step to reproduce:

1. I am from East Europe
2. Create a new KeyVault in East US
3. Run simplest terraform script
  
  ```
  terraform {
  
    backend "local" {
    }
  
    required_providers {
     
      azurerm = {
        source  = "hashicorp/azurerm"
        version = "~> 2.75.0"
      }
     
    }
  }
  provider "azurerm" {
    features {}
  }
  
  resource "azurerm_key_vault_secret" "test_key" {
    name         = "test"
    value        = "test"
    key_vault_id = "/subscriptions/<subscription_id>/resourceGroups/<rg-name>/providers/Microsoft.KeyVault/vaults/<vault_name>"
  }
  ```
4. Secret has been created
5. Terraform output is "Error: Provider produced inconsistent result after apply"

**Comment by sinbai** [comment]
@viper4u Per the log provided, I found the Azure keyVault cannot be found after setting the secret, Could you help confirm  whether the keyvault exists after reproducing this issue?

It is best to confirm through the following link：
https://docs.microsoft.com/en-us/rest/api/resources/resources/list#code-try-0

Notes:
1. Update the API version with 2020-06-01.
2. Add the optional parameter `$filter` with value `resourceType eq 'Microsoft.KeyVault/vaults' and name eq '<kv name>'` like this:
![getResource](https://user-images.githubusercontent.com/39109137/132684446-4731ca90-7427-43ba-bd25-59487cc6d477.png)
3. Try it several times to see if you can find it every time.

**Comment by viper4u** [comment]
Results
1) when I use the filter with "and name eq '<kv name>'", then I get 
`{
  "value": [],
  "nextLink": "<some_url>"
}`
2) when I use the filter without "and name eq '<kv name>'", then I get list of resources with my KeyVault
3) When I try another Keyvault in another subscription then I get result without nextlink
`
{
  "value": [
    {
      "id": "/subscriptions/<my_subscription>/resourceGroups/<rg_name>/providers/Microsoft.KeyVault/vaults/<kv_name>",
      "name": "makctestkv3",
      "type": "Microsoft.KeyVault/vaults",
      "location": "eastus2"
    }
  ]
}
`

**Comment by sinbai** [comment]
> Results
> 
> 1. when I use the filter with "and name eq ''", then I get
>    `{   "value": [],   "nextLink": "<some_url>" }`
> 2. when I use the filter without "and name eq ''", then I get list of resources with my KeyVault
> 3. When I try another Keyvault in another subscription then I get result without nextlink
>    `{   "value": [ { "id": "/subscriptions/<my_subscription>/resourceGroups/<rg_name>/providers/Microsoft.KeyVault/vaults/<kv_name>", "name": "makctestkv3", "type": "Microsoft.KeyVault/vaults", "location": "eastus2" }   ] }`

Could you please share with me the screenshot of the above results and obscure sensitive information?

**Comment by sinbai** [comment]
I submitted a [PR](https://github.com/hashicorp/terraform-provider-azurerm/pull/13409) that might solve this issue. Since I can’t reproduce this problem until now, I can’t verify it. Is anyone willing to build it and verify that this issue is fixed? The info to build a TF provider (on Windows) are as follows:

**Requirements:**
Terraform version 0.12.x + (but 1.x is recommended)
Go version 1.16.x (to build the provider plugin)
Git Bash for Windows
Make for Windows

**Using the locally compiled Azure Provider binary:**
 For example, add the following to terraform.rc for a provider binary located in D:\GoCode\bin. The file named terraform.rc and placed in the relevant user's %APPDATA% directory.
![image](https://user-images.githubusercontent.com/39109137/133888703-48a1a7ef-ba20-4e8e-916a-e65acf24ef0f.png)

**Steps:**
1. Pull this [PR](https://github.com/hashicorp/terraform-provider-azurerm/pull/13409) 
2. Run `make build` in git bash.

After `make build` is completed, it will be used automatically when running `terraform apply`.

**References:**
https://github.com/hashicorp/terraform-provider-azurerm/blob/main/README.md

**Comment by viper4u** [comment]
@sinbai  WE have tried your PR #13409 and it solves our problem

**Comment by tplive** [comment]
@sinbai I'm trying to confirm if this solved the problem, so I've created a mock tf setup, and I'm deploying to the same subscription using the same SP as where I had the problem, but running only the creation of the vault secret that fails in production. I'm getting the same error:

**Comment by sinbai** [comment]
> @sinbai We have tried your PR #13409 and it solves our problem

Thank you for your kind cooperation.

**Comment by sinbai** [comment]
> However, I'm new to running developer versions of the provider, so I'm unsure if I did it correctly. It's still downloading the azurerm 2.78 version, even though I'm warned that I'm using developer override. Is it supposed to still dl the prod version? 🤷🏽

@tplive  It will be downloaded if you run `terraform init` when using provider development overrides. if so, please skip `terraform init`, It is not necessary and may error unexpectedly. 

BTW, The PR #13409  depends on the trace log and confirmation information provided by viper4u. The same error maybe caused by different reasons, so, if the tf config and trace log (after applying PR #13409) can be provided, I would try to find the reason for the failure.

**Comment by tplive** [comment]
@sinbai thank you for the advice. I've stripped it down further, so the azurerm provider is the only thing remaining, it deploys only a single kv secret now, and still fails.. :) I can send you the config and trace log if you like to see what that looks like?

**Comment by sinbai** [comment]
@tplive Please send the config and trace log to sinbai@hotmail.com, thank you.

**Comment by dma-sitecore** [comment]
@tombuildsstuff @sinbai  The response from Azure Support
1.	The primary recommendation to resolve this issue is to move to Azure Resource Graph instead of the List Resources API. You can ignore all of the API implementation details and nextLink challenges and just make a single call to Search-AzGraph. When the portal displays a list of resources, it is done using Azure Resource Graph rather than the List Resources API. You can replace all of your code with just this one line:

Search-AzGraph -Query "resources | where type =~ 'Microsoft.KeyVault/vaults' and name =~ '$AzureKeyVaultName'" -Subscription $SubscriptionId

**Comment by sinbai** [comment]
@dma-sitecore Thank you for the information. The [PR ](https://github.com/hashicorp/terraform-provider-azurerm/pull/13719)of moving to Resource Graph API instead of Resource API when retrieving Keyvault is ready.

 Is anyone willing to build it and verify that this issue can be fixed with this PR because I can‘t repro it?

**Comment by alwalkerTH** [comment]
Using the provider built from that PR solved the issue for me. I had a manually created Key Vault in another subscription entirely and every time I added a secret to it it would fail with the "root resource was present, but now absent" error and then subsequent applies would complain that the secret already existed.

I don't know if it's related but I also couldn't import the existing secret either.  It was created but any time I tried to do terraform import it would complain that it didn't, yet when I ran apply it complained that it did.

**Comment by tplive** [comment]
(sorry I first posted this on the PR, then realized I should have been posted it here instead :) )

@sinbai I have now tested this PR with our environment. I add four secrets in deployment, and for testing I made sure two of them already existed. The first apply will report that it should create four secrets. Then will apparently deploy two secrets correctly, and correctly report that the two other secrets already exist and have to be imported into state.

However, when I run the same deployment a second time, it reports that it will again create four secrets, and says all four already exist. So the two secrets were deployed on the first run, without error, but were not added to the state, as far as I can tell.

**Comment by sinbai** [comment]
> > @tplive Thank you very much. I want to clarify that this [PR ](https://github.com/hashicorp/terraform-provider-azurerm/pull/13719) only applies to the errors "root resource was present, but now absent", and the fix depends on the trace log provided. So, could you help confirm that this PR can indeed fix that error? Thanks again for all your help.
> 
> I can confirm that the "root resource was present, but now absent" was not found in the trace log.

@tplive As far as I know, this error happens intermittently. So, can I consider that this PR can fix it, right?

**Comment by tplive** [comment]
> @tplive As far as I know, this error happens intermittently. So, can I consider that this PR can fix it, right?

For us, the error happened every time. But if the fix is to remove the error message, then it does the job. :)

**Comment by aniro** [comment]
We are facing the same issue on:
hashicorp/azurerm v2.83.0
Terraform v1.0.9

**Comment by tplive** [comment]
> We are facing the same issue on:
> hashicorp/azurerm v2.83.0
> Terraform v1.0.9
> 
> 

Interested in hearing more about your particular situation, as we are also still suffering this bug. 😊

**Comment by aniro** [comment]
@tplive 
> Interested in hearing more about your particular situation, as we are also still suffering this bug. 😊

It is similar to the situations already described above. 
Keyvault is first created and populated in scope of terraform configuration `step_1` without issues.
Then the same keyvault is referenced in configuration `step_2` by its id using `data.azurerm_key_vault.keyvault_name.id`. `step_2` fails most of the time (but not always) with the following error:
```
Error: Provider produced inconsistent result after apply
```

**Comment by tplive** [comment]
> @tplive and @aniro, is there any chance that the module used to create the KeyVault is not the in the same subscription as the one used to create the secret (just as @Aldiuser mentioned)?

@ magodo In our case, the the Vault has been previously created and exists in the same state, all in the same subscription.

**Comment by magodo** [comment]
@tplive I see. Then the root cause might be due to the resource list API issue, though I'm not certain sure. However, the PR #14047 should still fix this case hopefully.

**Comment by aniro** [comment]
> @tplive and @aniro, is there any chance that the module used to create the KeyVault is not the in the same subscription as the one used to create the secret (just as @Aldiuser mentioned)?

@magodo both steps are configured in a similar way and use the same subscription in our case.

**Comment by magodo** [comment]
@tplive, @aniro do you mind to do a test on my PR in your environment where the issue occurs?

**Comment by tplive** [comment]
@magodo I will have time to test this today, hopefully to completion. I will report back the result asap.

**Comment by tplive** [comment]
@magodo  I have run through the test as done previously, and I can confirm that this appears to have solved the issue for us! 👍🏽

**Comment by cspring86** [comment]
We're seeing a similar issue.

We've pinned the issue down to a call made by the TF provider to the Azure REST API. This one:

GET https://management.azure.com/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resources?$filter=resourceType eq 'Microsoft.KeyVault/vaults' and name eq 'this-is-my-kv'&$top=5&api-version=2020-06-01

Depending on **where** the Terraform script is run from (i.e. the IP of the machine that the TF script is run on), it either works or doesn't work.

Example:
- If the runner and Key Vault locations are the same (e.g. eastus), it works.
- If the runner and Key Vault locations are different (e.g. eastus/uksouth), it doesn't work.

Another thing to mention is that when using the Azure REST API to directly fetch the Key Vault by ID, it works. It's purely this resource list filter call that has the issue.

Hope that helps resolve the issue.

**Comment by cspring86** [comment]
The only alternative I can think of is that there's some kind of replication issue between regions.

If I remove the `name eq 'this-is-my-kv'` from the Azure REST API query, I do in fact get Key Vaults from eastus, even though I'm in the UK, but I don't see all of them - I don't see the most recently created Key Vaults.

**Comment by yube-sitecore** [comment]
Hello Everyone,
Thank you everyone for your effort on this issue.
It gave a lot of pain to us as well.

I just checked the freshest image, which is '1.1.0-beta2' and all my terraform deployments passed successfully, after over 20 failed attempts while using the 1.0.6 version. 

I will test it a few days more and back to you with some feedback.

**Comment by yube-sitecore** [comment]
Unfortunately - no, the problem still be very stable.

**Comment by mybayern1974** [comment]
@yube-sitecore , thank you for proactively updating here! In the meantime, would you mind provide your repro steps as detail as you can to serve us who are still tracking this issue including raising tickets to Microsoft KeyVault as well as ARM team? 

Below are some mentioned repro steps but unfortunately we are unable to follow any of them:
1. Programmatically create a new Azure subscription and right after that create KV and KV Sec. (Given I do not have an environment to programmatically create a subs so have not tried this yet)
2. Use TF to create KV in region A (say WestUS) and then use TF to create KV Secret that depends on that KV in region B (say EastUS). This repro step is introduced in [issue 14249](https://github.com/hashicorp/terraform-provider-azurerm/issues/14249). However, though we can follow that repro steps, we were still unable to repro.

Lastly, would that be feasible for you to provide correlationIDs during all your TF calls? With getting those ID the Azure ARM team might be able to track what happened in the backend. No worry if you are not sure how to get that and we are happy to share instructions accordingly.

**Comment by mybayern1974** [comment]
@cspring86 thank you for updating your story and glad to see you are unblocked! 

While on the other hand, I may feel there are other people who may still feel blocked/inconvenient by this case, because as Terraform users, people may expect things to happen automatically and consistently, so for people who have repro steps that can trigger this issue and those steps need to happen regularly and automatically per some business needs, manually filing Azure support ticket might not be a solution to them.

I'm still trying to contact Azure support to get their API usage best practice, with that we may wanna accordingly tune Terraform implementation to introduce a clean refinement.

**Comment by yube-sitecore** [comment]
> @yube-sitecore , thank you for proactively updating here! In the meantime, would you mind provide your repro steps as detail as you can to serve us who are still tracking this issue including raising tickets to Microsoft KeyVault as well as ARM team?
> 
> Below are some mentioned repro steps but unfortunately we are unable to follow any of them:
> 
> 1. Programmatically create a new Azure subscription and right after that create KV and KV Sec. (Given I do not have an environment to programmatically create a subs so have not tried this yet)
> 2. Use TF to create KV in region A (say WestUS) and then use TF to create KV Secret that depends on that KV in region B (say EastUS). This repro step is introduced in [issue 14249](https://github.com/hashicorp/terraform-provider-azurerm/issues/14249). However, though we can follow that repro steps, we were still unable to repro.
> 
> Lastly, would that be feasible for you to provide correlationIDs during all your TF calls? With getting those ID the Azure ARM team might be able to track what happened in the backend. No worry if you are not sure how to get that and we are happy to share instructions accordingly.

Hello and thank you for your effort,
We are doing a lot of deployments from scratch every day using the same set of complex terraform scripts.
The issue happens pretty often - it can work a few days, but after fails for 2-3 weeks during each deployment because of that issue.

I see that Maksym has provided a sample previously, https://github.com/hashicorp/terraform-provider-azurerm/issues/11059#issuecomment-915798838 . That's our case.

I see that we have opened a ticket to MS, but there is no clear details - because the ticket was closed - as can't reproduce.

I tried to find some patterns with a time of the day\locations\etc - no, it's totally random for us.
I will try to find more details from guys who worked on this issue previously and will back to you.

Also if any logs can be helpful - I am ready to collect and give them for the troubleshooting needs.

**Comment by cspring86** [comment]
> @cspring86 thank you for updating your story and glad to see you are unblocked!
> 
> While on the other hand, I may feel there are other people who may still feel blocked/inconvenient by this case, because as Terraform users, people may expect things to happen automatically and consistently, so for people who have repro steps that can trigger this issue and those steps need to happen regularly and automatically per some business needs, manually filing Azure support ticket might not be a solution to them.
> 
> I'm still trying to contact Azure support to get their API usage best practice, with that we may wanna accordingly tune Terraform implementation to introduce a clean refinement.

@mybayern1974 I 100% agree. To your point, and @yube-sitecore 's by the looks of it too, we're still seeing this issue as well and it's very frustrating. Like @yube-sitecore , we can go days without hitting the issue and then I've hit the issue again today - same pipelines, no changes.

Based upon the response I got from Azure Support, my only guess is that there's some persistent issues with ARM caching at the moment, either longer than expected delays or just never being updated, possibly due to some high load on Azure (similar to what we saw at the start of the pandemic).

**Comment by yube-sitecore** [comment]
> > @cspring86 thank you for updating your story and glad to see you are unblocked!
> > While on the other hand, I may feel there are other people who may still feel blocked/inconvenient by this case, because as Terraform users, people may expect things to happen automatically and consistently, so for people who have repro steps that can trigger this issue and those steps need to happen regularly and automatically per some business needs, manually filing Azure support ticket might not be a solution to them.
> > I'm still trying to contact Azure support to get their API usage best practice, with that we may wanna accordingly tune Terraform implementation to introduce a clean refinement.
> 
> @mybayern1974 I 100% agree. To your point, and @yube-sitecore 's by the looks of it too, we're still seeing this issue as well and it's very frustrating. Like @yube-sitecore , we can go days without hitting the issue and then I've hit the issue again today - same pipelines, no changes.
> 
> Based upon the response I got from Azure Support, my only guess is that there are some persistent issues with ARM caching at the moment, either longer than expected delays or just never being updated, possibly due to some high load on Azure (similar to what we saw at the start of the pandemic).

We face the same for quite a long time. It's a big blocker for us. 
I plan to start an investigation of this issue from the scratch, as we need to find a workaround whenever it will be.

**Comment by mybayern1974** [comment]
> > @cspring86 thank you for updating your story and glad to see you are unblocked!
> > While on the other hand, I may feel there are other people who may still feel blocked/inconvenient by this case, because as Terraform users, people may expect things to happen automatically and consistently, so for people who have repro steps that can trigger this issue and those steps need to happen regularly and automatically per some business needs, manually filing Azure support ticket might not be a solution to them.
> > I'm still trying to contact Azure support to get their API usage best practice, with that we may wanna accordingly tune Terraform implementation to introduce a clean refinement.
> 
> @mybayern1974 I 100% agree. To your point, and @yube-sitecore 's by the looks of it too, we're still seeing this issue as well and it's very frustrating. Like @yube-sitecore , we can go days without hitting the issue and then I've hit the issue again today - same pipelines, no changes.
> 
> Based upon the response I got from Azure Support, my only guess is that there's some persistent issues with ARM caching at the moment, either longer than expected delays or just never being updated, possibly due to some high load on Azure (similar to what we saw at the start of the pandemic).

@cspring86 may I possibly get your contact, say email address or whatever, though I'm not sure the best practice to share that for respecting privacy. The purpose is I wanna get detail of your saying "Filing an azure support ticket could unblock things once,  though later still repro", specially, that ticket number or Azure support folk contact. With getting that, I plan to possibly reach out to some Azure internal service team to help clarify a concern that I intuitively have: If manual efforts can flush to make things work, why cannot that magic happen automatically/by-default?

**Comment by techy321** [comment]
Terraform version v0.13.4
Azurerm provider 2.88.1

Following and hoping for a fix. We are experiencing the same issue with our pipeline sporadically. The keyvault gets created during an earlier stage in the pipeline and these secrets get created during a later stage in the pipeline. The secrets appear in the portal but not in state.

Error: Provider produced inconsistent result after apply

When applying changes to module.SQL.azurerm_key_vault_secret.sql_password,
provider "registry.terraform.io/hashicorp/azurerm" produced an unexpected new
value: Root resource was present, but now absent.

**Comment by techy321** [comment]
> Terraform version v0.13.4 Azurerm provider 2.88.1
> 
> Following and hoping for a fix. We are experiencing the same issue with our pipeline sporadically. The keyvault gets created during an earlier stage in the pipeline and these secrets get created during a later stage in the pipeline. The secrets appear in the portal but not in state.
> 
> Error: Provider produced inconsistent result after apply
> 
> When applying changes to module.SQL.azurerm_key_vault_secret.sql_password, provider "registry.terraform.io/hashicorp/azurerm" produced an unexpected new value: Root resource was present, but now absent.

I've also noticed during times when this occurs, we are also not able to complete an az keyvault list either even though the keyvault is shown in the portal. Once cli can list keyvault again, we are able to re-run pipeline successfully.

**Comment by techy321** [comment]
> > Terraform version v0.13.4 Azurerm provider 2.88.1
> > Following and hoping for a fix. We are experiencing the same issue with our pipeline sporadically. The keyvault gets created during an earlier stage in the pipeline and these secrets get created during a later stage in the pipeline. The secrets appear in the portal but not in state.
> > Error: Provider produced inconsistent result after apply
> > When applying changes to module.SQL.azurerm_key_vault_secret.sql_password, provider "registry.terraform.io/hashicorp/azurerm" produced an unexpected new value: Root resource was present, but now absent.
> 
> I've also noticed during times when this occurs, we are also not able to complete an az keyvault list either even though the keyvault is shown in the portal. Once cli can list keyvault again, we are able to re-run pipeline successfully.

Breaking discovery!!! If you add a tag to the keyvault, prior to running tf it will be able to successfully find the kv.

**Comment by yube-sitecore** [comment]
> > > Terraform version v0.13.4 Azurerm provider 2.88.1
> > > Following and hoping for a fix. We are experiencing the same issue with our pipeline sporadically. The keyvault gets created during an earlier stage in the pipeline and these secrets get created during a later stage in the pipeline. The secrets appear in the portal but not in state.
> > > Error: Provider produced inconsistent result after apply
> > > When applying changes to module.SQL.azurerm_key_vault_secret.sql_password, provider "registry.terraform.io/hashicorp/azurerm" produced an unexpected new value: Root resource was present, but now absent.
> > 
> > 
> > I've also noticed during times when this occurs, we are also not able to complete an az keyvault list either even though the keyvault is shown in the portal. Once cli can list keyvault again, we are able to re-run pipeline successfully.
> 
> Breaking discovery!!! If you add a tag to the keyvault, prior to running tf it will be able to successfully find the kv.

Hello @techy321 ,
Could you share details, please - what tag you have added and in what way?

**Comment by techy321** [comment]
> > > > Terraform version v0.13.4 Azurerm provider 2.88.1
> > > > Following and hoping for a fix. We are experiencing the same issue with our pipeline sporadically. The keyvault gets created during an earlier stage in the pipeline and these secrets get created during a later stage in the pipeline. The secrets appear in the portal but not in state.
> > > > Error: Provider produced inconsistent result after apply
> > > > When applying changes to module.SQL.azurerm_key_vault_secret.sql_password, provider "registry.terraform.io/hashicorp/azurerm" produced an unexpected new value: Root resource was present, but now absent.
> > > 
> > > 
> > > I've also noticed during times when this occurs, we are also not able to complete an az keyvault list either even though the keyvault is shown in the portal. Once cli can list keyvault again, we are able to re-run pipeline successfully.
> > 
> > 
> > Breaking discovery!!! If you add a tag to the keyvault, prior to running tf it will be able to successfully find the kv.
> 
> Hello @techy321 , Could you share details, please - what tag you have added and in what way?

I just added a tag with any value to the key vault in question, immediately after the cli was able to return the keyvault if i do a list, tf as well.

**Comment by yube-sitecore** [comment]
> 

Thank you for sharing - I will test this workaround from our end and will write an update for other interested people if it works for us as well.

**Comment by asebetich** [comment]
Any updates on this :)? 

I am creating an "azurerm_synapse_linked_service" and specifically am trying to create a connection to a Key Vault and also an SFTP server. The resource "azurerm_synapse_linked_service" is looped using the for_each command. 

This looped resource creates the keyvault Linked Service then uses that linked service to create the SFTP linked service (stores password details in the key vault). 

When running the plan, i get all green check marks! However, running the apply provideded me with the following error: 
Provider produced inconsistent result after apply
value: Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.


Any help is appreciated! I am going to try and dissect the for_each and create one at a time to see if i get a different result

**Comment by yube-sitecore** [comment]
Hello @asebetich ,
Welcome to the club =)

We have the issue when we create secrets and after reading them within the same script run.
As a solution - in case of failure - we implementing retryer which will import resources into the state. 
The idea that import into state is also not stable in this scenario, and as someone found - adding a tag to the resources doing some effect - we don't know what and how it affects, but the fact that if in case of failure we will add any tag (as described above) to resource, import existing resources into state and execute it again - execution will pass.

**Comment by asebetich** [comment]
@yube-sitecore, hmm. I might not fully understand your suggestion here. 

So in my specific case, I use a data call to grab some secrets for my Azure Synapse Workspace (sql username/pw). I do not create the Key vault in my main.tf file, i just reference it and the associated secrets i need with the data resource block. 

That being said, when the Synapse resource is already created, I can't seem to create these linked services. I removed my SFTP Linked Service and just tried to create one using the Key Vault connection and still nothing. 

So what resource are you suggesting I add a tag to?

**Comment by yube-sitecore** [comment]
I agree with you @asebetich , that our cases are a bit different.
But the root cause seems the same. 

The issue that during the planning phase - azure reporting that secret exists.
But when terraform requests it - It can't be returned and you see this error message.

There are many ideas why it's happening, we spend months with Microsoft trying to get the root cause and find that it can be related to broken paging implementation within the endpoint which used by terraform. Later another guy shared their experience with Microsoft support and it was different.

Here we discuss how at least to workaround the problem, as the problem is still actual and not resolved between terraform and Azure.

In our case workaround:

1. Catch this exact error
2. Add tag with any value (like test:test) to KeyVault resource
3. Run terraform import command to import explicitly existing secrets which already exist, and which terraform can't obtain during apply phase.
4. Execute apply again and it will pass

Otherwise, it can be unpredictable how many times you will be required to execute it to bypass.
Another finding from our end is that in most cases issues occur when KeyVault is deployed within a different location from the place where you execute your script. Like KeyVault deployed in West Europe, but you machine\agent\doesn't how you run your terraform executed from Australia - almost in all cases this execution wouldn't pass from the first attempt. But at the same time - no guarantee that it will pass successfully from first attempt if those locations will be collocated.

**Comment by asebetich** [comment]
@yube-sitecore Thanks for the quick responses! I see what you mean now with literally the workaround being to add a tag? haha very odd behavior for that to work. 

I initially packaged all the synapse workspace related items into 1 module for my own organization, and thought that separating the linked_service creation to not be included into the module might somehow fix this issue, but alas, nothing changed and I keep getting the same error we are all facing. 

I'm going to continue to troubleshoot this and will now try to see if literally manually adding a tag to the Key Vault helps. All my resources are in the same region, so that part doesn't apply to me. I'll keep this thread updated with my findings (if i have any)

**Comment by yube-sitecore** [comment]
Yes, @asebetich  add tag and import secret data into state explicitly.

**Comment by michaelelleby** [comment]
Importing into state explicitly is very impractical for production deployments, as the person deploying usually will not have permissions to modify the state for production. This needs a real fix.

**Comment by yube-sitecore** [comment]
@michaelelleby yes, agree. But we face this issue mostly during the initial provisioning time.

We were not able to resolve this issue with Microsoft.
I will consider moving from KeyVault to Hashicorp Vault potentially, as this issue was an open a year ago.

**Comment by magodo** [comment]
@cspring86 Just want to echo @mybayern1974's ask, is it possible to provide some hint (e.g. the support name, id, etc) about the support ticket that manually resolved your issue as you mentioned [above](https://github.com/hashicorp/terraform-provider-azurerm/issues/11059#issuecomment-982421493)? So that we can internally find the support guy to learn details, and hopefully figure out some solution to resolve this issue.

**Comment by magodo** [comment]
Another update is that I can (intermittently) reproduce this with following steps:

1. Check out and build one of my personal branch for the provider: https://github.com/magodo/terraform-provider-azurerm/tree/endpoint ([diff against the main branch](https://github.com/hashicorp/terraform-provider-azurerm/compare/main...magodo:endpoint)). The only diff is that this branch added an environment variable: `ARM_RESOURCE_MANAGER_ENDPOINT_PREFIX`, which allows users to target to a (region) specific ARM endpoint
2. Use above provider to provision the [example template](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/key_vault_secret), which the `azurerm_key_vault_secret` commented out, and `ARM_RESOURCE_MANAGER_ENDPOINT_PREFIX` set to `eastus`
3. Uncomment the `azurerm_key_vault_secret` part and run `terraform apply` again, but this time with `ARM_RESOURCE_MANAGER_ENDPOINT_PREFIX` set to soemthing different than `eastus`, e.g. `norwayeast`

I ran above steps for 3 times, and the issue occurs 2 times.

So another way to get it fixed is to ensure *the regional ARM endpoint used to create the key vault nested resources is the same as the **resource group***, which contains the key vault (based on this [comment](https://github.com/Azure/AKS/issues/1964#issuecomment-729150191)). There is one issue #13364 that asks for adding the regional endpoint support for the provider, just FYI.

**Comment by RavulaChetan** [comment]
Hi
I am seeing the same error with keyvault at adding secrets.
Do we have a solution on this?

**Comment by magodo** [comment]
@RavulaChetan Not really, only a workaround: https://github.com/hashicorp/terraform-provider-azurerm/issues/11059#issuecomment-1018661692

**Comment by RavulaChetan** [comment]
Thanks @magodo; Is this work around for keyvault already existing and we add secrets?
What about if the keyVault is created in the same code as adding secrets?

Also when the work around says add Tag to the KeyVault; can you elaborate more please.

**Comment by Azrrael** [comment]
Hi everyone. We run into this issue a few weeks ago, while using Terraform 1.0.0 with azurerm 2.96.0. We tried the tagging (both to the actual secret and the Vault itself) solution and we also requested a ARM cache sync to Azure support, but neither worked.
When trying to import manually the values manually to our .tfstate, we got an error pointing out the said values didn't exist, contradicting the results of an apply. Also and as an FYI, the Vault in question has purge protection enabled.

**Comment by magodo** [comment]
@RavulaChetan The workaround is to add a regular resource tag in the key vault via other means (e.g. portal, azure-cli), and optionally remove the tag. This might (no proof) trigger a refresh of the ARM cache in the target regional endpoint, which causes your next `GET` call on its `/resources` endpoint returns the key vault. As @Azrrael stated, they found the workaround above doesn't work either, which means this is not a reliable workaround..

For a long term solution, we might have following solutions:
1. Allows the provider to set ARM regional endpoint, which is tracked in https://github.com/hashicorp/terraform-provider-azurerm/issues/15632
2. Change the resource id of the key vault nested resources, which is tracked in #16174

**Comment by vijaytdh** [comment]
In case this helps anyone else but I received this error (`Root resource was present, but now absent.` and the error `Cannot import non-existent remote object` when attempting to import) too but in my case it was my own silly mistake because in my code I had multiple providers defined for different subscriptions.

When adding the secret I was passing the key vault resource id via a datasource that used the right provider, however in the module that creates the secret I was not specifying the provider so it was defaulting to the subscription for the default provider - which in my case is not where the key vault resides for the secret I was adding - I only noticed it when enabling the trace logging to see the actual API request that was made.

**Comment by Divya1388** [comment]
Facing the same error when writing terratests for SQL FOG which includes creation of keyvault and CMK Key.
And it is very consuming to trougleshoot this generic error and left without any clue now.
Have tried adding time_sleep resource to wait after the keyvault creation and also tried specifying providers where the script is creating keys and secrets. None of them worked.
Will try adding tag to the keyvault

**Comment by meilz381** [comment]
I had also had the error "Root resource was present, but now absent.". The key vault is in another subscription (created by a different setup). Maybe someone else helps it, I was able to solve my issue by putting the key vault resources in an extra module and setting the provider config there with the other subscription id.

**Comment by Azrrael** [comment]
Hi there. We run into this issue again. We were using a homemade module to create a DB and users, saving two passwords, passwordA and passwordB on Key Vault. We have multiple environments and in particular, Dev and QA share a common Key Vault, called Lower. We used **the exact same code** for all environments, changing only parameters such as environment name, DB size, etc. When we run our code, it worked in all environments except Dev. So, we have two calls with the same code, saving passwords on the same key vault, with one working and the other failing. Furthermore, we tried "run apply, then create a tag on the secret, then run import" and this time it worked, but not on the first try but a couple of days later.
Not sure why is this happening, but here's hoping this info helps somebody.

**Comment by brisitw** [comment]
We are facing the same issue when creating secrets via `azurerm_key_vault_secret` with terraform azurerm provider 3.5.0.

**Comment by o1da** [comment]
Hi guys,
I had exactly the same issue too, with azurerm provider 2.99.0 and 3.9.0. 

I was passing ID of a keyvault directly as a string, then I wonder maybe I am passing wrong value and tried to read keyvault with data source and pass ID parameter from there. Then I realized I can't read that keyvault as I am not on correct subscription...

So `az account set -s ...` or rather setting subscription with the provider did the trick for me.

Pretty misleading error message I would say :)

**Comment by RobertFloor** [comment]
We have the same problem, we provide the azurerm provider to a module. The module then should sync keys from a key vault in a different subscription. However, somehow terraform believes the key is not there and proceeds to delete them from the state. It then tries to re add them but they are still there and this gives an error


```
module "test" {
  resource-group-kv = "test1"
 key-vault-name = "test2"
  providers = {
    azurerm = azurerm.cicd
  }
}


provider "azurerm" {
  features {
    key_vault {
      purge_soft_delete_on_destroy = true
    }
  }
  alias           = "cicd"
  subscription_id = "xxxxx"
}
```

In the module


```
data "azurerm_key_vault" "azvault" {
  name                = var.key-vault-name
  resource_group_name = var.resource-group-kv
}

resource "azurerm_key_vault_secret" "secret-developer-write" {
  name         = "super-secret"
  value        =  "secret"
  key_vault_id = data.azurerm_key_vault.azvault.id
}
```

**Comment by RobertFloor** [comment]
In our case the root case was this issue : https://github.com/hashicorp/terraform-provider-azurerm/issues/8680 (Azurerm providers in module will use az account active subscription instead of the configured credentials). We could solve the problem by including the provider both in the calling code and in the module:

We do get this warning, but it still solves the problem and terraform loads the state correctly, including a key vault from a different subscription. 


╷```

│ Warning: Redundant empty provider block
│
│   on ../../modules/kafka-topic/backend.tf line 17:
│   17: provider "azurerm" {
│
│ Earlier versions of Terraform used empty provider blocks ("proxy provider configurations") for child modules to declare their need to be passed a provider configuration by
│ their callers. That approach was ambiguous and is now deprecated.
│
│ If you control this module, you can migrate to the new declaration syntax by removing all of the empty provider "azurerm" blocks and then adding or updating an entry like
│ the following to the required_providers block of module.kafka-topics:
│     azurerm = {
│       source = "hashicorp/azurerm"
│       configuration_aliases = [
│         azurerm.subscription-config,
│       ]
│     }
```

**Comment by MrBasset** [comment]
Just to document another scenario where this problem appears to occur, similar to others here.

We have a central Key Vault in which we store shared secrets. We were using data sources to pull back account details that were then used to create TLS certs that are then loaded into a subscription specific key vault. In our case, we missed explicitly adding the provider to the data source (which still worked), but that somehow caused the wrong subscription ID to be used when refreshing the state on the Key Vault (or secrets within it).

**Comment by magodo** [comment]
👏 #17407 has just been merged. Appreaciate anyone suffering from this issue can test whether that solves your issue in the upcoming new release v3.12.0, or via a local build using the `main` branch!

**Comment by elongstreet88** [comment]
Still seeing terrible performance for keyvault secrets and the perpetual value:[]; nextlink: "..." with this release as mentioned above.
I found https://github.com/hashicorp/terraform-provider-azurerm/pull/13409 was rejected, however, testing it locally, it speeds up our deploys nearly 5x:

Without https://github.com/hashicorp/terraform-provider-azurerm/pull/13409
- terraform apply  71.60s user 24.44s system 31% cpu 5:00.74 total

With
- terraform apply  71.57s user 23.69s system 108% cpu 1:27.57 total

While i think there are a few different issues going on at the same time here, could we re-consider 13409 for a quick fix?
I'm happy to PR it again if needed.
I think this is uniquely affecting people with have hundreds of keyvaults in their subscriptions like us, and that $top logic on the filter wrecks us the more we add.

Thanks.

**Comment by Tbohunek** [comment]
@magodo I just tested and sadly it doesn't work for me with `3.12.0` or I do something wrong. ☹️

**Comment by magodo** [comment]
@Tbohunek Can you provide more details about your use case, i.e. the config and the log.

**Comment by Tbohunek** [comment]
@magodo what log exactly? I've updated the above config.

Nothing fancy: Vault is in another subscription, so even if I use the correct `provider` for `data azurerm_key_vault`, I still need to use that `provider` also for all stored values, be it `resource` or `data`.

For other resource types I'm used to it being able to work out from the `.id` which includes `/subscriptions/sss` and hence not needing a `provider` to select the subscription.

This "problem" prevents some use-cases such as passing external `var.vault_id` into the `module`, because when you change to vault in another subscription, it breaks. This is actually how I "found" this issue in the first place.

**Comment by magodo** [comment]
@Tbohunek Yes, in cross subscription use, the "other resource types" works because their id has the subscription id in it. While as the id of the key vault nested items are using the data plane id, which doesn't have the sub id. So when reading them, the provider assumes the subscription is the current provider's subscription.

I'd like to confirm that the config in https://github.com/hashicorp/terraform-provider-azurerm/issues/11059#issuecomment-1172423670 works with specifying the provider alias.

**Comment by Tbohunek** [comment]
@magodo I see. But why does it then search for it in a resource group instead of using the data-plane ID which is globally unique? And where does it get the resource group from, if not from the `resource_id`? It's weird.
But yes, specifying the `provider` in both resources does the trick.

**Comment by magodo** [comment]
The `key_vault_id` in the `azurerm_key_vault_secret` data source is the control plane resource id. From this resource id, the provider will invoke the `GET` on the key vault to get its data plane URL, see [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/6257739685f7729e9af8ce5e059d74e2d06f3ad3/internal/services/keyvault/key_vault_key_data_source.go#L133) and [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/6257739685f7729e9af8ce5e059d74e2d06f3ad3/internal/services/keyvault/client/helpers.go#L52).

**Comment by Tbohunek** [comment]
@magodo I cannot agree, because the error message reveals the resource group name, which I do not give to it. It has to come from the id.

**Comment by magodo** [comment]
@Tbohunek Indeed, whilst currently the provider aims to work in a subscription level. The client of each service is initialized with a fixed subscription (the same as the provider): https://github.com/hashicorp/terraform-provider-azurerm/blob/6257739685f7729e9af8ce5e059d74e2d06f3ad3/internal/services/keyvault/client/client.go#L23. Some of the reason is the limitation of the underlying Azure SDK used, where the methods didn't expose the subscription id as a parameter. However, this might be changed after we switch to use https://github.com/hashicorp/go-azure-sdk.

**Comment by t-l-k** [comment]
I had providers described like this:

```
provider "azurerm" {
  subscription_id = module.estate.subscription.id
  features {}
}

provider "azurerm" {
  features {}
  alias           = "hub"
  subscription_id = module.estate.hub_subscription.id
}
```

And was attempting to create a certificate like this, which worked for the create:

```
resource "azurerm_key_vault_certificate" "winrm" {
  ...
}
``` 

But failed on subsequent plan operations.

I was expecting it to use the unaliased provider with `subscription_id` sourced from, in my case, `module.estate.subscription.id` but it was actually trying with the `subscription_id` assigned to the `hub` aliased provider, as judged by `TF_LOG = Debug`, much to my surprise. This resulted in the `ResourceGroupNotFound` 404 error already reported.

I tried:

```
resource "azurerm_key_vault_certificate" "winrm" {
  provider = azurerm
  ...
}
```

But it still failed, so I added another provider as a repeat of my unaliased provider:

```
provider "azurerm" {
  subscription_id = module.estate.subscription.id
  alias           = "automations"
  features {}
}
```

And reconfigured the certificate:

```
resource "azurerm_key_vault_certificate" "winrm" {
  provider = azurerm.automations
  ...
}
```

And it worked, although now I have two providers for essentially the same ARM endpoint? I don't necessarily want to alias my default local provider though as conceptually it relates to a known configuration tier, and it means more lines of boilerplate. 

It's surprisingly inconsistent, please fix 🙏🙏

**Comment by Nemeczek** [comment]
Lmao. I have the same issue. 
`provider "provider[\"registry.terraform.io/hashicorp/azurerm\"]" produced an unexpected new value: Root resource was present, but now absent.`

I am simply saving generated token to KV as a secret. 

```
resource "azurerm_key_vault_secret" "git_devops_pat_dbw_secret" {
  name            = "git-devops-dbw-pat-${var.dbw_name}"
  value           = databricks_token.pat.token_value
  key_vault_id    = data.azurerm_key_vault.tokens_kv.id
  expiration_date = time_rotating.azure.rotation_rfc3339
}
```

Terraform v1.2.2 on windows_amd64
+ provider registry.terraform.io/databricks/databricks v1.2.1
+ provider registry.terraform.io/hashicorp/azurerm v3.20.0
+ provider registry.terraform.io/hashicorp/random v3.3.2
+ provider registry.terraform.io/hashicorp/time v0.8.0

**Comment by L-u-k-e-GIT** [comment]
@Nemeczek do you have Key vault on a different subscription where you are creating resource. In this call, use module and change provider when calling it

**Comment by Nemeczek** [comment]
> @Nemeczek do you have Key vault on a different subscription where you are creating resource. In this call, use module and change provider when calling it

I just did that and it worked. It is literally copy paste from main module (I am using Azure CLI context as auth)
to the module that needs to create secret. 

```
provider "azurerm" {
  features {}
  skip_provider_registration = true
}
```

**Comment by yube-sitecore** [comment]
Look at this thread, about the KeyVault issue

**Comment by magodo** [comment]
@Nemeczek Could you please try out my [local branch](https://github.com/magodo/terraform-provider-azurerm/tree/key_vault_arg) of the provider, which changes the way of how to *determine the Resource ID for the Key Vault at URL*? To use that branch, please checkout it locally, and install it to somewhere. Then follow [this page](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/DEVELOPER.md#developer-using-the-locally-compiled-azure-provider-binary) to setup terraform to using the locally built provider.

If that works, then please let us know.

**Comment by pearcec** [comment]
FWIW, we experienced this a lot, we have taken to this workaround, this gets added to all our automation. Azure's response has been use Graph API.  It seems that the eastus is the primary for Keyvault lookups since it is a global resource (not 100% sure).  It seems to work regardless of the region.  

```
 sudo echo "$(host eastus.management.azure.com | tail -1 | awk '{ print $NF }') management.azure.com" | sudo tee -a /etc/hosts
```

**Comment by rvplauborg** [comment]
I just experienced this issue, where the workaround of adding an aliased provider with same subscription as my default provider, and adding that provider to key vault and key vault secrets resources fixed the issue.

**Comment by llohrstraeter** [comment]
Having the same issue. I use a module with multiple providers that contains two more modules. One of them is provisioning a keyvault in subscription A, the other one looks up a keyvault using a datasource from subscription B.

When removing the module that fetches the keyvault from sub B, the deployment of keyvault to sub A works fine.
When adding the module fetching the keyvault from sub B, deployment of keyvault to sub A fails because it tries to fetch the resource group from sub B instead of sub A.
The strange thing is that I don't even pass the provider for sub B to the module that deploys the keyvault to sub A. No clue why it is checking the wrong one.

Even though the apply fails the keyvault is deployed properly to sub A and I get an error that it already exists when running another apply.

Using TF version 1.3.6 with azurerm V3.34.0, happy to provide trace logs if needed.

**Comment by rezarms** [comment]
I upgrade azurerm to 3.41 and fixed the issue

**Comment by vamshisiram** [comment]
I am getting the same issue with azurerm_service_plan resource.

I use a module to deploy a bunch of apps, everything works fine expect the app service plan resource fails with this error.

if I run a plan/apply again it says a resource with same name already exists. When I refresh the azure portal, I see it. I deleted the app service plan from portal and retry this again, it fails with same error. It is a loop.

It appears as if terraform is not able to reach azure api in time to see the created resource.

I use three different providers in the module.

I am using terraform 1.3.9 and azurerm version 3.45.0.

it worked fine until few days ago with version 3.2.0 and it stopped working all of a sudden. I opened a support case with hashicorp and their recommendation is to use version latest of azurerm. So, I did that and it still does not work.

pls help with this issue.

**Comment by bricksterlad** [comment]
I was getting the same error with azurerm version 3.13.0, then I moves to 3.45.0 and I managed to create the resource without getting an error. But now when I run a terraform apply the second time, it tries to create the same resource a second time , so I get an error that the resource already exists. When I checked the state file, I couldn't find the resource there after the second apply (it was there after the first successful apply)
I would appreciate any help regarding this

**Comment by akorp** [comment]
Thanks everyone for info here in this issue. We also had this error:
```
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to
│ module.<redacted>.azurerm_key_vault_secret.service_principal_password,
│ provider "provider[\"registry.terraform.io/hashicorp/azurerm\"]" produced
│ an unexpected new value: Root resource was present, but now absent.
│ 
│ This is a bug in the provider, which should be reported in the provider's
│ own issue tracker.
╵
```
In our case we use two providers for 2 different subscription - `default` and `second`. We got reference to a key vault in the second subscription with specifying provider.
```
data "azurerm_key_vault" "kv_1" {
  provider = azurerm.second

  name                = kv_name
  resource_group_name = rg_name
}
```
This key vault ID was later used in a module to create `resource "azurerm_key_vault_secret"` where we didn't specify any provider so a default was used.

The fix was to pass provider to the module, so the `second` provider is used inside the module to create `resource "azurerm_key_vault_secret"` 
```
module "service_principal" {
  source = "url"

  providers = {
    azurerm = azurerm.gpb
  }

...
```

**Comment by rmcolbert** [comment]
We are experiencing the same kind of issue with the azurerm_app_service_certificate resource. We're passing in the fully qualified certificate secret ID (the only resource input related to the KV) but since the API to create call to Microsoft.Web/certificates requires the KV's resource ID, the provider has to try and resolve the resource ID. This is a big issue for us because the App Service Certificate will always be in a different subscription from the KV.

In our case, if the azurerm_app_service_certificate accepted either kv_cert_secret_id or {kv id + cert name}, we could get around this ... or just a way to pass in the kv subscription id for it to use in the id resolution.

**Comment by nadworny** [comment]
> I was getting the same error with azurerm version 3.13.0, then I moves to 3.45.0 and I managed to create the resource without getting an error. But now when I run a terraform apply the second time, it tries to create the same resource a second time , so I get an error that the resource already exists. When I checked the state file, I couldn't find the resource there after the second apply (it was there after the first successful apply) I would appreciate any help regarding this

same here, when running it for the second time, I get the "already exists" error. Using azurerm 3.105.0

**Comment by tombuildsstuff** [comment]
👋 hey folks

We've recently updated the behaviour of the Key Vault cache such that we check both the Key Vaults List API and the same (old) Resources API that the current version of Azure CLI is currently using - and as such it'd be good to hear if this is still being encountered when using the latest version of the Provider? This fix is included in `v3.106.0` and above (although `v3.107.0` is the latest at the time of writing) - if you're able to would you mind updating to the latest version of the Provider and letting us know if that works for you?

Thanks!

**Comment by serbezki** [comment]
I just hit this problem with version 3.108 of the provider.

**Comment by nadworny** [comment]
I tried it with 3.108 and it still doesn't work

**Comment by DevopsMercenary** [comment]
I've been experiencing this issue for a while.. Here's my current state
Terraform v1.9.4
on darwin_arm64
+ provider registry.terraform.io/hashicorp/azuread v2.53.1
+ provider registry.terraform.io/hashicorp/azurerm v3.116.0
+ provider registry.terraform.io/hashicorp/time v0.12.0

**Comment by DevopsMercenary** [comment]
In my case I am using a time_rotating provider to set the expiration_date

```
resource "time_rotating" "avd_registration_expiration" {
  # Must be between 1 hour and 30 days
  rotation_days = 29
}

resource "azurerm_virtual_desktop_host_pool_registration_info" "default" {

  hostpool_id     = data.azurerm_virtual_desktop_host_pool.default.id
  expiration_date = time_rotating.avd_registration_expiration.rotation_rfc3339

}

resource "azurerm_key_vault_secret" "registration-key" {

  name = "${data.azurerm_virtual_desktop_host_pool.default.name}-registration-key"

  value = azurerm_virtual_desktop_host_pool_registration_info.default.token

  key_vault_id = data.azurerm_key_vault.default.id

  expiration_date = time_rotating.avd_registration_expiration.rotation_rfc3339


}


```

**Comment by fy-openai** [comment]
we're still seeing this with v3.109.0. A similar one is mentioned here: https://github.com/hashicorp/terraform-provider-azurerm/issues/24948. And a potential fix was made here: https://github.com/hashicorp/go-azure-sdk/pull/927 not for terraform-provider-azurerm, but for pandora.

The root cause IMO is nextLink is not read(unmarshalled) properly from http repsonse, effectively causing the ExecutePaged function to [return prematurely](https://github.com/hashicorp/go-azure-sdk/blob/main/sdk/client/client.go#L618), leading [ListBySubscriptionComplete](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/keyvault/client/populate_cache.go#L46C41-L46C67) to return 10(default size of a list call) keyvaults instead of the full list. The same thing applies to [ListComplete](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/keyvault/client/populate_cache.go#L72), which has a bigger default size: 24.

**Comment by teowa** [comment]
Hi @fy-openai , the custom pager ( [x-ms-pageable.nextLinkName](https://github.com/Azure/azure-rest-api-specs/blob/74faa8fdb1043d2c45f1f3ae1e7808a38c1535ee/specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2023-07-01/keyvault.json#L398-L400) ) was handled in hashicorp/go-azure-sdk PR https://github.com/hashicorp/go-azure-sdk/pull/1039 on Jun 28, and this was introduced to Azurerm at PR https://github.com/hashicorp/terraform-provider-azurerm/pull/26502 , released at [3.111.0 version (Jul 4)](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/CHANGELOG-v3.md#31110-july-04-2024) , so after this version the issue should be resolved.

**Comment by trevorcreech** [comment]
Confirmed fixed for me on 3.111.0

**Comment by DevopsMercenary** [comment]
Ran into this issue again.  Current versions include

```
Terraform v1.9.8
on darwin_arm64
+ provider registry.terraform.io/hashicorp/azuread v3.0.2
+ provider registry.terraform.io/hashicorp/azurerm v4.7.0
+ provider registry.terraform.io/hashicorp/time v0.12.1
```

**Comment by mnapoleon** [comment]
I have run into this issue as well when attempt to add a virtual endpoint to an existing Postgres Flexible server.
Terraform v1.9.7
on darwin_arm64
+ provider registry.terraform.io/azure/azapi v2.0.1
+ provider registry.terraform.io/hashicorp/azuread v2.53.1
+ provider registry.terraform.io/hashicorp/azurerm v3.116.0
+ provider registry.terraform.io/hashicorp/random v3.6.3

`Error: Provider produced inconsistent result after apply
│
│ When applying changes to
│ azurerm_postgresql_flexible_server_virtual_endpoint.***-postgres-server-vnet-endpoint[0],
│ provider "provider[\"registry.terraform.io/hashicorp/azurerm\"]" produced
│ an unexpected new value: Root object was present, but now absent.
│
│ This is a bug in the provider, which should be reported in the provider's
│ own issue tracker.`

**Comment by nathanblair** [comment]
This remains an issue on AzureRM 4.18

**Comment by holograph** [comment]
Also 4.37.0 😢

**Comment by dmpe** [comment]
Yes, can confirm that. Currently facing this with 4.38.1

**Comment by katbyte** [comment]
Thank you for the report, and apologies for the slow follow up here.

This issue is linked to #17407 (`azurerm_key_vault` data source: cache the `vault_uri` once read), which merged and shipped in v3.12.0 and appears to have fixed it, so we are closing this one out as resolved.

If you are still running into this on the current (v5.x) version of the provider, please let us know with a comment or open a new issue referencing this one and we will take another look.
